src/wayland-eglsurface.c

```diff
@@ -1291,19 +1293,25 @@ acquire_surface_image(WlEglDisplay *display, WlEglSurface *surface)
         }
     }
 
-    for (i = 0; i < surface->ctx.numStreamImages; i++) {
-        if (surface->ctx.streamImages[i]->eglImage == eglImage) {
-            image = surface->ctx.streamImages[i];
-            break;
+    pthread_mutex_lock(&surface->ctx.streamImagesMutex);
+
+    // Locate the corresponding WlEglStreamImage
+    {
+        EGLBoolean found = EGL_FALSE;
+
+        wl_list_for_each(image, &surface->ctx.streamImages, link) {
+            if (image->eglImage == eglImage) {
+                found = EGL_TRUE;
+                break;
+            }
         }
-    }
 
-    if (!image) {
-        goto fail_destroy_sync;
+        if (!found) {
+            pthread_mutex_unlock(&surface->ctx.streamImagesMutex);
```

Comment:
I think it would make sense to move this `pthread_mutex_unlock` into `fail_destroy_sync`. This matches how we unlock the mutex in `fail_release` and also prevents us from accidentally forgetting to unlock it if we use `fail_destroy_sync` from a second location in the future.

Reply:
I believe jumping to fail_release actually makes more sense since the eglImage here is already acquired. It also looks like fail_release needed some changes for how image is handled now.